manual/src/cmds/intf-c.etex

```diff
@@ -2652,6 +2652,10 @@ CAMLprim stub_gethostbyname(value vname)
   struct hostent * h;
   char * name;
 
+  /* Check that the OCaml string doesn't contain any embedded null
+     bytes */
+  if (! caml_string_is_c_safe(name)) caml_raise_not_found();
```

Comment:
I know that's what `Unix.gethostbyname` does but I don't think it's a good idea to conflate this error condition with a valid response from the function. That's the kind of stuff that eventually gets you the usability disaster that `Sys.file_exists` is.

I think this should be either of:

1. Raise `Invalid_argument`. That means you expect the client to check the string for null bytes before calling the function. Maybe not a good idea since it's never going to do that.
2. Raise `Failure "Stray null byte in name"`. That means your API will need to document another exception besides `Unix.Unix_error`. That's ok if it's the only `Failure` raise of the function. A little bit less if it's not the only one as it becomes difficult to get a processing path for this error condition which one may want for this case.
3. Raise `Unix.Unix_error` with `EINVAL`. That means you simulate a response from the OS. It's potentially confusing but that's ok in my book. IIRC the `Unix` module is not shy from doing so when it makes sense.

I would suggest to do 3 in this example (in turn it would also show the support the OCaml C interface has for `Unix_error` in `unixsupport.h` which could also perhaps be documented at some point). Also I'm not sure which C style is followed but I think in general there's no space between ! and identifiers in C.

So this wall of text to suggest perhaps:
```
if (!caml_string_is_c_safe(name)) caml_unix_error (EINVAL, "gethostbyname", name);
```

Reply:
> I would suggest to do 3 in this example (in turn it would also show the support the OCaml C interface has for Unix_error in unixsupport.h which could also perhaps be documented at some point).

Good points, I'll go with the 3rd option. Maybe a comment such as `/* return an error or raise an exception */` would have sufficed.

> Also I'm not sure which C style is followed but I think in general there's no space between ! and identifiers in C.

The rationale is often that it makes it clearer that `!` is needed (it's written this way in the `gethost.c` code). I often get confused by the `!` OCaml de dereference operator...

and I've just noticed that the function is missing a return type ;)

Reply:
> Maybe a comment such as `/* return an error or raise an exception */` would have sufficed.

I think it's better to give users a good solution to cut and paste. Rather than leave them to think about that in a hurry and do the wrong thing.

> I often get confused by the `!` OCaml de dereference operator...

My favourite line of code in my executables is always:

```ocaml
let () = if !Sys.interactive then () else exit (main ())
```
Because it even manages to confuse myself :-)